docs/proposals/ktl.md

```diff
@@ -0,0 +1,243 @@
+- build new cli commands in `kubernetes/kubectl` and vendor them into `kurnetes/ktl`
+- build common cli Kubernetes client infrastructure and libraries that can be used to develop a decentralized cli ecosystem
+
+### ktl
+
+In a new repo (`kubernetes/ktl`), create the `ktl` binary that dispatches (both statically and dynamically)
+to commands developed in other repos.
+
+#### Dispatch
+
+Static dispatch:
+
+- vendor in kubernetes/kubernetes/cmd/kubectl cobra commands, and register them under `ktl kube`
+- vendor in kubernetes/kubectl/cmd cobra commands, and register them under `ktl`
+- vendor in commands from other sources as needed over time
+
+Dynamic dispatch:
+
+- support git-style plugins that allow registering new commands under `ktl`
+  - use simplified version of the kubectl plugin implementation - make configuration files optional
+- plugins only purpose is discovery of kubernetes related commands
+  - plugins can leverage shared client cli libraries (whether they are installed as plugins or not)
+- by default throw an error if plugin names conflict with existing commands
+  - this is configurable
+- plugins can be disabled through modifying `~/.ktlconfig`
+
+Overriding existing commands:
+
+- support command alias' to allow overriding one command with another
+  - allows plugins to extend and override built-in commands
+  - allows `ktl kube *` commands to be aliased to `ktl *`
+
+#### Configuration
+
+Use [viper](https://github.com/spf13/viper) to configure dispatch precedence
+
+### Cli commands
+
+All new cli commands should be built outside of kubernetes/kubernetes and vendored into the kubernetes/ktl.  Command repos
+implement an interface that returns a list of cobra (sp13) commands, which ktl can register under the root.
+
+Initially new cli commands can be built in the kubernetes/kubectl, but if needed, development maybe moved to other repos.
+
+#### Conventions
+
+- ktl top level commands are one of
+  - command groups that work against a specific API group/versions (e.g. isto, svcat, kube)  e.g. `create`
+    for a given resource.
+  - generic commands that are agnostic to specific APIs, but instead discover API metadata and work against
+    all APIs following published conventions.  e.g. a `scale` that works against anything with a scale sub resource
+
+
+### Library infrastructure
+
+Develop shared set of client / cli libraries that handle registering flags / environment variables / config files
+and provide common functionality to cli clients of Kubernetes.
+
+- reading and parsing kubeconfig
+- printing libraries
+- reading and parsing config into objects from files, directories, stdin
+- indexing + querying discovery API and openapi data
+- manipulating unstructured objects using openapi schema
+- merging objects
+- generating patches from objects
+- writing objects to the server - optimistic locking failure retries with exp backoff
```

Comment:
This needs rethinking. If you're writing a client from scratch, you should assiduously avoid update.  The issues with skew, dropped fields, and update are real.  External clients provide no benefit or protection over internal types, so don't think you'll get away from it by using them.

Also, patch already retries for you server-side.

Reply:
Patch is broken.  There are a number of fields it doesn't work for.

Reply:
> Patch is broken. There are a number of fields it doesn't work for.

Focusing on that problem seems like it will bear more fruit than trying to resolve the skew issues that have plagued clients for many years with open, unstaffed proposals to fix them.

Reply:
There isn't a good way to fix it without breaking backward compatibility.  :(

Reply:
We can easily introduce new versioned patch mime types to fix issues.

If we don't fix patch, then we're basically saying users will continue to be broken - that has to have a really good justification.

Reply:
>  The issues with skew, dropped fields, and update are real.

If I understand correctly, these issues are caused by serialization using go-structs.  Using unstructured doesn't have this issue, and IIUC proto3 will soon also keep fields when skewed.

> Also, patch already retries for you server-side.

This isn't sufficient in cases where the value being written is based off the value read.  An example is incrementing a counter depends on the value of the counter that was read.  Client-side retries on optimistic locking failures are a well understood pattern and we should support it.

> Focusing on that problem seems like it will bear more fruit than trying to resolve the skew issues that have plagued clients for many years with open, unstaffed proposals to fix them.

What are the skew issues you are referring to besides the dropped fields?  As I understand it, the problem with dropping fields is specific to the serialization, not the API itself.

> If we don't fix patch, then we're basically saying users will continue to be broken - that has to have a really good justification.

Agreed.  We should make a decision and justify that decision regardless of what it is.  We haven't done a good job of supporting strategic merge PATCH, and we should be honest about whether we are going to do a better job in the future.  It would be helpful if there was a document that describes the original justification for creating our own bespoke PATCH format and what other options where considered instead, do you know of one?  In my experience it has created more problems than it has solved.

Regardless, the ownership for strategic merge PATCH should come out of api machinery.  Fixing that API should not fall upon the clients.

Reply:
> This isn't sufficient in cases where the value being written is based off the value read. An example is incrementing a counter depends on the value of the counter that was read. Client-side retries on optimistic locking failures are a well understood pattern and we should support it.

I'm still not seeing how that precludes using patch since you can include resourceVersion in those cases and still avoid dropped fields.

Reply:
2 other things to consider w.r.t. going all in on PATCH / strategic merge PATCH:

- My understanding is that not every resource type supports strategic merge patch
- PATCH only helps with dropping fields for compiled in types that are skewed, but does not help for either 1) extension types that aren't compiled in or 2) new types that aren't compiled into old clients (yet)

Reply:
> I'm still not seeing how that precludes using patch since you can include resourceVersion in those cases and still avoid dropped fields.

That doesn't (though there are a number of other reasons I would give not to use PATCH), however it does preclude relying exclusively on server-side retries for optimistic locking failures.  I believe that is what this discussion was originally about:

> writing objects to the server - optimistic locking failure retries with exp backoff